kernel/mem/virtmem.cpp

```diff
@@ -1,34 +1,134 @@
 /**
  * 
  */
+#include <kernel/device/uart.h>
+#include <kernel/kstdlib/kstdio.h>
+#include <kernel/mem/address.h>
 #include <kernel/mem/virtmem.h>
 
-namespace VirtMemoryManager
-{
+extern uint32_t __RAM_START;
+extern uint32_t __RAM_END;
+extern uint32_t __FREE_RAM_START;
+static constexpr uint32_t PHYS_PAGE_SIZE = 0x1000;
+static constexpr uint32_t L1_TABLE_ENTRY_DEFAULT = 0x00000001;
+static constexpr uint32_t L2_TABLE_ENTRY_DEFAULT = 0x00000032;
+static constexpr uint32_t DEVICES_MEMORY_BASE = 0x1c000000;
+static constexpr uint32_t GIC_MEMORY_BASE = 0x2c000000;
 
-static uint16_t L1_translation_table_index(VirtualAddress address)
+VirtMemoryManager::VirtMemoryManager()
 {
-    return address.get() >> 20;
+    /* First we build the translation table to simply map the kernel and devices */
+    L1TranslationTable* l1_table_base = allocate_L1_translation_table();
+    initialise_L1_translation_table(*l1_table_base);
+    identity_map_kernel(*l1_table_base);
+    map_devices(*l1_table_base);
+
+    /* Now we can use this table as our map and enable the MMU */
+    switch_L1_translation_table(l1_table_base);
+    CPU::invalidate_unified_TLB();
+    CPU::set_DACR_by_domain(Domain::ALL_DOMAINS, DACRValue::CLIENT);
+    CPU::set_SCTLR_flag(SCTLRFlag::MMU_ENABLE);
 }
 
-static uint8_t L2_translation_table_index(VirtualAddress address)
+void VirtMemoryManager::identity_map_kernel(L1TranslationTable& table)
 {
-    return (address.get() >> 12) & 0xFF;
+    L2TranslationTable* l2_table_base;
+    uint32_t free_ram_start = reinterpret_cast<uint32_t>(&__FREE_RAM_START);
+    uint32_t ram_start = reinterpret_cast<uint32_t>(&__RAM_START);
+    uint32_t kernel_pages = (free_ram_start - ram_start) / PHYS_PAGE_SIZE;
+    uint32_t current_page = ram_start;
+    uint16_t l1_offset_start = L1_translation_table_index(ram_start);
+    uint16_t l1_offset_end = l1_offset_start + (kernel_pages / ENTIRES_PER_L2_TRANSLATION_TABLE);
+
+    for(uint32_t i = l1_offset_start; i <= l1_offset_end; ++i)
+    {
+        l2_table_base = reinterpret_cast<L2TranslationTable*>(PhysicalMemoryManager::obj_instance().allocate_1k_page());
+        table.entries[i].entry = reinterpret_cast<uint32_t>(l2_table_base) | L1_TABLE_ENTRY_DEFAULT;
+
+        for(uint32_t j = 0; j < ENTIRES_PER_L2_TRANSLATION_TABLE && current_page < free_ram_start; ++j)
+        {
+            l2_table_base->entries[j].entry = current_page | L2_TABLE_ENTRY_DEFAULT;
+            current_page += PHYS_PAGE_SIZE;
+        }
```

Comment:
It might be better to have a function called `mapAddress(Phys, Virt)` instead of for loops. That way we can call `map` outside of the virtual memory manager (for example during `Process` creation)